### PR TITLE
Add ParticleNet to NanoAOD

### PR DESCRIPTION
**Pull request by hqucms**
#### PR description:

This PR adds the ParticleNet tagger to NanoAOD. The following scores will be included in NanoAOD:

- ParticleNet: `TvsQCD`, `WvsQCD`, `ZvsQCD`, `HbbvsQCD`, `HccvsQCD`, `H4qvsQCD`, `QCD`
- ParticleNet-MD: `Xbb`, `Xcc`, `Xqq`, `QCD`

The ParticleNet tagger needs to run in the NanoAOD sequence until we have the tagger stored in MiniAOD. The updated training [[V01](https://github.com/cms-sw/cmssw/pull/31036)] should be used for both UL and EOY samples.

FYI @camclean @alefisico

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by mariadalfonso** [comment]
This PR add 9 floats for a total of 37 floats of ID for the AK8 ? 
https://github.com/cms-sw/cmssw/blob/872f4c73e28af813940442d79ee81213ada26df5/PhysicsTools/NanoAOD/python/jets_cff.py#L408-L439
https://github.com/cms-sw/cmssw/blob/872f4c73e28af813940442d79ee81213ada26df5/PhysicsTools/NanoAOD/python/jets_cff.py#L398-403
Are the SF available for all these ?

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 1]
```diff
@@ -313,18 +317,21 @@ def nanoAOD_customizeCommon(process):
         nanoAOD_addDeepBTag_switch = cms.untracked.bool(False),
         nanoAOD_addDeepBoostedJet_switch = cms.untracked.bool(True), # will deactivate this in future miniAOD releases
         nanoAOD_addDeepDoubleX_switch = cms.untracked.bool(True), 
+        nanoAOD_addParticleNet_switch = cms.untracked.bool(True),
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:320]: do you need to re-run the particleNet also when is available in the mini ?
For the re-mini we can use the one stored in the miniAOD ?
For the old samples [ run2_nanoAOD_94X2016, run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2, run2_nanoAOD_102Xv1, run2_nanoAOD_106Xv1 ] indeed we need to re-run.

**Comment by hqucms** [comment]
@mariadalfonso 
SFs will be available for ParticleNet `TvsQCD`, `WvsQCD` and all the ParticleNet-MD scores. The long-term plan is to replace most of the DeepAK8 taggers by their ParticleNet counterparts, but depending on the production plans I guess the DeepAK8 taggers will still be needed by ongoing analyses.

File: PhysicsTools/NanoAOD/python/nano_cff.py (hunk 1 above)
Comment (hqucms) [reply PhysicsTools/NanoAOD/python/nano_cff.py:320]: > do you need to re-run the particleNet also when is available in the mini ?
No. We can use the ones stored in MiniAOD once they are available (e.g., in UL re-mini).

Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/nano_cff.py:320]: ok. good the taggers are already available in master version.

--> for master:
 nanoAOD_produceParticleNet_switch = cms.untracked.bool(False)
if run2_nanoAOD_106Xv1 + anyOld:
 nanoAOD_produceParticleNet_switch = cms.untracked.bool(True)
add deepParticleNet

for 10_6:
if NOT run2_miniAOD_devel:
 nanoAOD_produceParticleNet_switch = cms.untracked.bool(True)
add ParticleNet

probably same logic should be applied to all the previous tagger

Comment (hqucms) [reply PhysicsTools/NanoAOD/python/nano_cff.py:320]: > ok. good the taggers are already available in master version.
> 
> --> for master:
> nanoAOD_produceParticleNet_switch = cms.untracked.bool(False)
> if run2_nanoAOD_106Xv1 + anyOld:
> nanoAOD_produceParticleNet_switch = cms.untracked.bool(True)
> add deepParticleNet
> 
> for 10_6:
> if NOT run2_miniAOD_devel:
> nanoAOD_produceParticleNet_switch = cms.untracked.bool(True)
> add ParticleNet
> 
> probably same logic should be applied to all the previous tagger

Yes indeed. Shall I change this or would you prefer to do it centrally afterwards?

Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/nano_cff.py:320]: should be done here and not afterwards

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> ParticleNet: `TvsQCD`, `WvsQCD`, `ZvsQCD`, `HbbvsQCD`, `QCD`

this is apparently incomplete,  HccvsQCD and H4qvsQCD are missing in the PR description

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31096	1532c62 
- code changes are in line with the PR description
    - most changes are in nanoAOD outputs (in fatJetTable particleNet\* variables)
    - in miniAOD only slimmedJetsAK8 are changed in jet tag content: a new HccvsQCD discriminator is added (slimmedJetsAK8 now has 110 discriminants; some reduction better be planned at some point)
- jenkins tests pass and comparisons with the baseline show differences only in nanoAOD and slimmedJetsAK8

**Comment by mariadalfonso** [comment]
@hqucms 
I looked at the results of the tests
[](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-08-09-2300+c69aaa/38269/validateJR/all_OldVSNew_TTbar13TeV2018nanoEDMwf10824p0/)

I can only see a subset of those added: i.e. missing WvsQCD(should be needed on ttbar events) and also H4qvsQCD and  MD_Xqq.  Is this expected ?

Also clarify if you plan to do this in this https://github.com/cms-sw/cmssw/pull/31096#discussion_r467380092
in this PR or in a subsequent PR for all AK8 fatjets

**Comment by mariadalfonso** [comment]
please test 136.8521

File: PhysicsTools/NanoAOD/python/nano_cff.py (hunk 1 above)
Comment (hqucms) [reply PhysicsTools/NanoAOD/python/nano_cff.py]: > should be done here and not afterwards

Done in https://github.com/cms-sw/cmssw/pull/31096/commits/2c7b9ae502a856b3f6f75e3665c9f47e322441d4.

**Comment by mariadalfonso** [comment]
@ahinzmann 
please comment on how you want the taggers need to be re-runned for previous eras EOY UL, taking into account the update in Puppi tunes and SF and calibration available.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ahinzmann** [comment]
> @ahinzmann
> please comment on how you want the taggers need to be re-runned for previous eras EOY UL, taking into account the update in Puppi tunes and SF and calibration available.

For UL only PUPPI tune v14 is supported. For EOY only PUPPI tune v11 is supported.
@hqucms Did you have evidence that the UL+PUPPIv14 training works on EOY+PUPPIv11? Otherwise, we would need a dedicated training included for EOY+PUPPIv11 in CMSSW.

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 2]
```diff
@@ -311,20 +315,32 @@ def nanoAOD_customizeCommon(process):
...
         jecPayload = cms.untracked.string('AK8PFPuppi')
         )
     # deepAK8 should not run on 80X, that contains ak8PFJetsCHS jets
     run2_miniAOD_80XLegacy.toModify(nanoAOD_addDeepInfoAK8_switch,
-                                    nanoAOD_addDeepBTag_switch = cms.untracked.bool(True),
-                                    nanoAOD_addDeepBoostedJet_switch = cms.untracked.bool(False),
-                                    nanoAOD_addDeepDoubleX_switch = cms.untracked.bool(False),
-                                    jecPayload = cms.untracked.string('AK8PFchs'))
+                                    nanoAOD_addDeepBTag_switch = True,
+                                    jecPayload = 'AK8PFchs')
```
Comment (mariadalfonso) [inline PhysicsTools/NanoAOD/python/nano_cff.py:326]: we are using now jecPayload = 'AK8PFchs' , but is this correct for the AK8 jets  ? 
I think we are using Puppi for AK8, so should we using the jecPayload = 'AK8PFpuppi')

Comment (ahinzmann) [reply PhysicsTools/NanoAOD/python/nano_cff.py:326]: yes, in 80X we used AK8 CHS jets, not PUPPI yet.

Comment (mariadalfonso) [reply PhysicsTools/NanoAOD/python/nano_cff.py:326]: ok

**Comment by hqucms** [comment]
> For UL only PUPPI tune v14 is supported. For EOY only PUPPI tune v11 is supported.
> @hqucms Did you have evidence that the UL+PUPPIv14 training works on EOY+PUPPIv11? Otherwise, we would need a dedicated training included for EOY+PUPPIv11 in CMSSW.

Yes, the UL+Puppi v14 training performs similarly or better than the previous training on EOY+Puppi v11 samples as well (see comparisons in [1]).

[1] https://indico.cern.ch/event/939494/contributions/3947427/attachments/2081304/3495876/20.07.28_JMAR_ParticleNet_for_UL.pdf#page=19 (last page)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31096	2c7b9ae 
- updates since the last signature from reco https://github.com/cms-sw/cmssw/pull/31096#issuecomment-671632361 were in nano category PhysicsTools/NanoAOD/python/nano_cff.py
- jenkins tests pass with a similar set of differences as in the earlier test (at least in the miniAOD set)

**Comment by mariadalfonso** [comment]
> @mariadalfonso
> SFs will be available for ParticleNet `TvsQCD`, `WvsQCD` and all the ParticleNet-MD scores. The long-term plan is to replace most of the DeepAK8 taggers by their ParticleNet counterparts, but depending on the production plans I guess the DeepAK8 taggers will still be needed by ongoing analyses.
> 
> Also given that on average the number of high-pt AK8 jets in typical events (e.g., ttbar) is <<1, the overall space increase should be quite small.

Can you quote separately the increase in space in mini and nano ? ttbar events are fine.

**Comment by mariadalfonso** [comment]
please test workflow 10224.15,1325.6,1325.7,1325.8,1329.1,136.7722,136.7952,136.8521,1325.81,136.8522,1325.61

**Comment by santocch** [comment]
+1

**Comment by hqucms** [comment]
> Can you quote separately the increase in space in mini and nano ? ttbar events are fine.

@mariadalfonso

Increase in Mini is negligible (1 out of 110 b-tag discriminants per ak8 jet).
Increase in Nano is topology dependent -- for ttbar the increase is ~0.01kb/evt [1]. 

[1] before: http://hqu.web.cern.ch/hqu/dev/pr31096-pre.html
after: http://hqu.web.cern.ch/hqu/dev/pr31096-post.html

**Comment by mariadalfonso** [comment]
+xpog

This PR add the particleNet in nano EOY, UL and master.
For the DeepAK8 this PR deactivates the recomputation for ULmini and mini made with master since it's already available in miniAOD while it still recompute the recipe for the EOY.

**Comment by mariadalfonso** [comment]
@hqucms please consider making the backport

**Comment by qliphy** [comment]
+1

**Pull request merged**